server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertRowPlan.java

```diff
@@ -414,27 +414,31 @@ private void putValues(DataOutputStream outputStream) throws QueryProcessExcepti
         ReadWriteIOUtils.write((String) values[i], outputStream);
       } else {
         ReadWriteIOUtils.write(dataTypes[i], outputStream);
-        switch (dataTypes[i]) {
-          case BOOLEAN:
-            ReadWriteIOUtils.write((Boolean) values[i], outputStream);
-            break;
-          case INT32:
-            ReadWriteIOUtils.write((Integer) values[i], outputStream);
-            break;
-          case INT64:
-            ReadWriteIOUtils.write((Long) values[i], outputStream);
-            break;
-          case FLOAT:
-            ReadWriteIOUtils.write((Float) values[i], outputStream);
-            break;
-          case DOUBLE:
-            ReadWriteIOUtils.write((Double) values[i], outputStream);
-            break;
-          case TEXT:
-            ReadWriteIOUtils.write((Binary) values[i], outputStream);
-            break;
-          default:
-            throw new QueryProcessException("Unsupported data type:" + dataTypes[i]);
+        if (isNeedInferType) {
+          ReadWriteIOUtils.write(values[i].toString(), outputStream);
```

Comment:
Why add this? Can we just infer types once? It look like the type needs to be inferred again after the change, whether it is restarted or sent to other nodes?

Reply:
You can image this situation: at first the data type has been inferred (data type is double but value is string type), then when the plan needs to be serialized, it will throw the exception like "String can not cast to Double".

Reply:
This is an optimization that can be done on our standalone version as well, otherwise it would be a waste of time to have to refer it again after restarting from WAL. In that case, that doesn't have much to do with the PR. You can do this logic for now, but you need to make it complete. For example, implement the same logic for "private void putValues(ByteBuffer buffer)", plus serialize/deserialize UT for your case.